Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,308 @@
+  veblenWith (ω ^ ·)
+
+@[simp]
+theorem veblen_zero : veblen 0 = fun a ↦ ω ^ a := by
+  rw [veblen, veblenWith_zero]
+
+theorem veblen_zero_apply (o : Ordinal) : veblen 0 o = ω ^ o := by
+  rw [veblen_zero]
+
+theorem veblen_of_ne_zero (h : o ≠ 0) : veblen o = derivFamily fun x : Set.Iio o ↦ veblen x.1 :=
+  veblenWith_of_ne_zero _ h
+
+theorem isNormal_veblen (o : Ordinal) : IsNormal (veblen o) :=
+  isNormal_veblenWith isNormal_omega0_opow o
+
+theorem veblen_veblen_of_lt (h : a < b) (c : Ordinal) : veblen a (veblen b c) = veblen b c :=
+  veblenWith_veblenWith_of_lt isNormal_omega0_opow h c
+
+theorem veblen_succ (o : Ordinal) : veblen (Order.succ o) = deriv (veblen o) :=
+  veblenWith_succ isNormal_omega0_opow o
+
+theorem veblen_right_strictMono (o : Ordinal) : StrictMono (veblen o) :=
+  veblenWith_right_strictMono isNormal_omega0_opow o
+
+theorem veblen_right_monotone (o : Ordinal) : Monotone (veblen o) :=
+  (veblen_right_strictMono o).monotone
+
+@[simp]
+theorem veblen_lt_veblen_right_iff : veblen o a < veblen o b ↔ a < b :=
+  veblenWith_lt_veblenWith_right_iff isNormal_omega0_opow
+
+@[simp]
+theorem veblen_le_veblen_right_iff : veblen o a ≤ veblen o b ↔ a ≤ b :=
+  veblenWith_le_veblenWith_right_iff isNormal_omega0_opow
+
+@[simp]
+theorem veblen_inj : veblen o a = veblen o b ↔ a = b :=
+  veblenWith_inj isNormal_omega0_opow
+
+theorem right_le_veblen (a b : Ordinal) : b ≤ veblen a b :=
+  right_le_veblenWith isNormal_omega0_opow a b
+
+theorem veblen_left_monotone (o : Ordinal) : Monotone fun a ↦ veblen a o :=
+  veblenWith_left_monotone isNormal_omega0_opow o
+
+@[simp]
+theorem veblen_pos (a b : Ordinal) : 0 < veblen a b :=
+  veblenWith_pos isNormal_omega0_opow omega0_opow_zero_pos a b
+
+theorem veblen_zero_strictMono : StrictMono (veblen · 0) :=
+  veblenWith_zero_strictMono isNormal_omega0_opow omega0_opow_zero_pos
+
+theorem veblen_zero_monotone : Monotone (veblen · 0) :=
+  veblen_zero_strictMono.monotone
+
+@[simp]
+theorem veblen_zero_lt_iff : veblen a 0 < veblen b 0 ↔ a < b :=
+  veblen_zero_strictMono.lt_iff_lt
+
+@[simp]
+theorem veblen_zero_le_iff : veblen a 0 ≤ veblen b 0 ↔ a ≤ b :=
+  veblen_zero_strictMono.le_iff_le
+
+@[simp]
+theorem veblen_zero_inj : veblen a 0 = veblen b 0 ↔ a = b :=
+  veblen_zero_strictMono.injective.eq_iff
+
+theorem left_le_veblen (a b : Ordinal) : a ≤ veblen a b :=
+  left_le_veblenWith isNormal_omega0_opow omega0_opow_zero_pos a b
+
+theorem isNormal_veblen_zero : IsNormal (veblen · 0) :=
+  isNormal_veblenWith_zero isNormal_omega0_opow omega0_opow_zero_pos
+
+theorem veblen_lt_veblen_iff :
+    veblen a b < veblen c d ↔ a = c ∧ b < d ∨ a < c ∧ b < veblen c d ∨ c < a ∧ veblen a b < d :=
+  veblenWith_lt_veblenWith_iff isNormal_omega0_opow
+
+theorem veblen_le_veblen_iff :
+    veblen a b ≤ veblen c d ↔ a = c ∧ b ≤ d ∨ a < c ∧ b ≤ veblen c d ∨ c < a ∧ veblen a b ≤ d :=
+  veblenWith_le_veblenWith_iff isNormal_omega0_opow
+
+theorem veblen_eq_veblen_iff :
```

Comment:
I find these theorems a bit hard to read. Could you add docstrings for them, describing the conclusion in words?